File: day27/StringCompression.java
public class StringCompressor {

    public static String compressString(String s) {
        if (s == null || s.isEmpty()) {
            System.out.println("Original: \"\"");
            System.out.println("Compressed: \"\"");
            return s;
        }

        StringBuilder compressed = new StringBuilder();
        int count = 1;

        for (int i = 1; i < s.length(); i++) {
            if (s.charAt(i) == s.charAt(i - 1)) {
                count++;
            } else {
                compressed.append(s.charAt(i - 1)).append(count);
                count = 1;
            }
        }

        compressed.append(s.charAt(s.length() - 1)).append(count);

        String result = compressed.toString();
        System.out.println("Original: " + s);
        System.out.println("Compressed: " + (result.length() < s.length() ? result : s));

        return result.length() < s.length() ? result : s;
    }

    public static void main(String[] args) {
        compressString("aabcccccaaa");
        compressString("abc");
    }
}

File: day27/string_compression.rs
fn compress_string(s: &str) -> String {
    if s.is_empty() {
        println!("Original: \"\"");
        println!("Compressed: \"\"");
        return String::new();
    }

    let mut compressed = String::new();
    let mut count = 1;
    let chars: Vec<char> = s.chars().collect();

    for i in 1..chars.len() {
        if chars[i] == chars[i - 1] {
            count += 1;
        } else {
            compressed.push(chars[i - 1]);
            compressed.push_str(&count.to_string());
            count = 1;
        }
    }

    compressed.push(*chars.last().unwrap());
    compressed.push_str(&count.to_string());

    println!("Original: {}", s);
    println!("Compressed: {}", if compressed.len() < s.len() { &compressed } else { s });

    if compressed.len() < s.len() {
        compressed
    } else {
        s.to_string()
    }
}

fn main() {
    compress_string("aabcccccaaa");
    compress_string("abc");
}
